src/effects/VST3/internal/x11/RunLoop.cpp

```diff
@@ -0,0 +1,119 @@
+/**********************************************************************
+
+  Audacity: A Digital Audio Editor
+
+  @file RunLoop.cpp
+
+  @author Vitaly Sverchinsky
+
+  @brief Part of Audacity VST3 module
+
+**********************************************************************/
+
+#include "RunLoop.h"
+
+using namespace internal::x11;
+
+IMPLEMENT_FUNKNOWN_METHODS(RunLoop, Steinberg::Linux::IRunLoop, Steinberg::Linux::IRunLoop::iid);
+
+gboolean RunLoop::OnEvent(GIOChannel *source, GIOCondition condition, gpointer data)
+{
+   auto handler = reinterpret_cast<Steinberg::Linux::IEventHandler*>(data);
+   handler->onFDIsSet(g_io_channel_unix_get_fd(source));
+   return G_SOURCE_REMOVE;
+}
+
+gboolean RunLoop::OnTimer(gpointer data)
+{
+   auto handler = reinterpret_cast<Steinberg::Linux::ITimerHandler*>(data);
+   handler->onTimer();
+   return G_SOURCE_CONTINUE;
+}
+
+RunLoop::RunLoop(::Display* display)
+   : mDisplay(display)
+{
+   FUNKNOWN_CTOR;
+}
+
+RunLoop::~RunLoop()
+{
+   FUNKNOWN_DTOR;
+}
+
+Steinberg::tresult RunLoop::registerEventHandler(
+   Steinberg::Linux::IEventHandler* handler, 
+   Steinberg::Linux::FileDescriptor fd)
+{
+   using namespace Steinberg;
+
+   if(handler == nullptr)
+      return kInvalidArgument;
+    
+   if(XConnectionNumber(mDisplay) == fd)
+      return kResultOk;
+
+   if(auto channel = g_io_channel_unix_new(fd))
+   {
+      auto sourceId = g_io_add_watch(
+         channel, 
+         (GIOCondition)(G_IO_IN | G_IO_OUT | G_IO_ERR | G_IO_HUP),
+         &RunLoop::OnEvent, 
+         handler
+      );
+      mEventHandlers[handler] = { channel, sourceId };
```

Comment:
The map is keyed on the handler only, not the handler and file descriptor pair.  Is that all right?  Or is it possible to reassign `mEventHandlers[handler]` ?